ansible_navigator/image_manager/image_puller.py

```diff
@@ -0,0 +1,158 @@
+            )
+            self._image_present = True
+
+        except subprocess.CalledProcessError as exc:
+            self._image_present = False
+            if "no such image" not in str(exc.stderr).lower():
+                message = "Image inspection failed, image assumed to be corrupted or missing"
+                self._log_message(level=logging.WARNING, message=message)
+
+    def _determine_pull(self):
+        if self._pull_policy == "missing" and self._image_present is False:
+            pull = True
+        elif self._pull_policy == "always":
+            pull = True
+        elif self._pull_policy == "tag" and self._image_tag == "latest":
+            pull = True
+        elif self._pull_policy == "tag" and self._image_present is False:
+            pull = True
+        else:
+            pull = False
+        self._pull_required = pull
+        self._assessment.pull_required = pull
+
+    def _extract_tag(self):
+        _image, _, image_tag = self._image.partition(":")
+        self._image_tag = image_tag or "latest"
+        message = f"Image tag is: {self._image_tag}"
+        self._log_message(level=logging.INFO, message=message)
+
+    def _log_message(self, message: str, level: int, hint=False):
+        if level == logging.ERROR:
+            self._exit_messages.append(ExitMessage(message=message))
+        elif hint:
+            self._exit_messages.append(ExitMessage(message=message, prefix=ExitPrefix.HINT))
+        else:
+
+            self._messages.append(LogMessage(level=level, message=message))
+        self._logger.log(level=level, msg=message)
+
+    def prologue_stdout(self):
+        """print a little value added information"""
+        messages = [("Execution environment image name:", self._image)]
+        messages.append(("Execution environment image tag:", self._image_tag))
+        messages.append(("Execution environment pull policy:", self._pull_policy))
+        messages.append(("Execution environment pull needed:", self._pull_required))
+        width = max((len(m[0]) + len(str(m[1])) + 2 for m in messages))
+        print("\u002d" * width)
+        print("Execution environment image and pull policy overview")
+        print("\u002d" * width)
+        column_width = max((len(m[0]) for m in messages))
+        for msg, value in messages:
+            print(f"{msg.ljust(column_width)} {value}")
+            self._log_message(message=f"{msg}: {value}", level=logging.INFO)
+        print("\u002d" * width)
+        print("Updating the execution enviroment")
+        print("\u002d" * width)
+
+    def pull_stdout(self):
+        """pull the image, print to stdout
+
+        podman writes to stderr
+        docker writes to stdout
+        """
+        try:
+            if self._container_engine == "podman":
+                subprocess.run([self._container_engine, "pull", self._image], check=True)
+            elif self._container_engine == "docker":
+                subprocess.run(
```

Comment:
Can we pipe both `stdout` ans `stderr` for both docker and podam and display stdout to stderr on console?

Reply:
I wanted to grab real steer where I could to format in our formet

Reply:
For docker, the progress bars are returned as stdout, and errors as stderr, so we can let stdout be printed on the screen, and capture stderr, render it with our [ERROR] format and log it.

For podman, the progress bars and errors are returned as stderr, so we have no choice but to let it print on the screen because there is no way to know what is just a message or error.

So this was the best case scenario, I suspect most users will be using docker, which in this scenario "looks" the best because we can color the errors, separate from the progress bars.